docs/proposals/04-new-core-gardener-cloud-apis.md

```diff
@@ -0,0 +1,698 @@
+
+---
+apiVersion: core.gardener.cloud/v1alpha1
+kind: BackupBucket
+metadata:
+  name: providerprofile1-random[:5]
+  - kind: ProviderProfile
+  ownerReferences:
+    name: providerprofile1
+spec:
+  provider:
+    type: <some-provider-name> # {aws,azure,gcp,...}
+    region: europe-central-1
+  secretRef:
+    name: backup-operator-provider
+    namespace: backup-garden
+status:
+  lastOperation:
+    description: Backup bucket has been successfully reconciled.
+    lastUpdateTime: '2020-04-13T14:34:27Z'
+    progress: 100
+    state: Succeeded
+    type: Reconcile
+  observedGeneration: 1
+  seed: seed1
+```
+
+### `BackupEntry` resource
+
+Special note: This new resource is cluster-scoped.
+
+```yaml
+# See also: https://github.com/gardener/gardener/blob/master/docs/proposals/02-backupinfra.md.
+
+apiVersion: v1
+kind: Secret
+metadata:
+  name: backup-operator-provider
+  namespace: backup-garden
+type: Opaque
+data:
+  # <some-provider-specific data keys>
+  # https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-alicloud/example/30-backupbucket.yaml#L9-L11
+  # https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-aws/example/30-backupbucket.yaml#L9-L10
+  # https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-azure/example/30-backupbucket.yaml#L9-L10
+  # https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-gcp/example/30-backupbucket.yaml#L9
+  # https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-openstack/example/30-backupbucket.yaml#L9-L13
+
+---
+apiVersion: core.gardener.cloud/v1alpha1
+kind: BackupEntry
+metadata:
+  name: shoot--core--crazy-botany--3ef42
+  namespace: garden-core
+  ownerReferences:
+  - apiVersion: core.gardener.cloud/v1alpha1
+    blockOwnerDeletion: false
+    controller: true
+    kind: Shoot
+    name: crazy-botany
+    uid: 19a9538b-5058-11e9-b5a6-5e696cab3bc8
+spec:
+  provider:
+    type: <some-provider-name> # {aws,azure,gcp,...}
+    region: europe-central-1
+  secretRef:
+    name: backup-operator-provider
+    namespace: backup-garden
+  shootUID: 19a9538b-5058-11e9-b5a6-5e696cab3bc8
+status:
+  lastOperation:
+    description: Backup entry has been successfully reconciled.
+    lastUpdateTime: '2020-04-13T14:34:27Z'
+    progress: 100
+    state: Succeeded
+    type: Reconcile
+  observedGeneration: 1
+  seed: seed1
+```
+
+### `Shoot` resource
+
+Special notes:
+
+* Moved `kubeProxy` and `kubelet` configuration to worker pools.
+* Moved remaining control plane configuration to new `.spec.provider.controlplane` section.
+
+```yaml
+apiVersion: core.gardener.cloud/v1alpha1
+kind: Shoot
+metadata:
+  name: crazy-botany
+  namespace: garden-core
+spec:
+  secretBindingName: secretbinding1
+# seedName: seed1
+  provider:
+    type: <some-provider-name> # {aws,azure,gcp,...}
+    profileName: providerprofile1
```

Comment:
As I mentioned above - the API is not clear - is `ProviderProfile` used for Pollicy and/or Mutation?

It's better for the `Shoot` resource to not know about the `ProviderProfile`

Reply:
> the API is not clear - is ProviderProfile used for Pollicy and/or Mutation?

Simple answer: both. What exactly is not clear?

> It's better for the Shoot resource to not know about the ProviderProfile

That won't be possible, I guess. We have introduced the `Cloud/ProviderProfiles` over a year ago because we needed some place where to put a) possible values for certain fields (e.g., machine types, machine images, whatever), and b) global configurations for this environment (e.g., the openstack keystone URL or request timeouts or whatever). Both is highly provider specific, and without a reference we cannot establish a correlation.

How should a solution look like?

Reply:
> How should a solution look like?

I would use `shootSelector` in the `ProviderProfile` to match Shoots.
This would allow Gardener cluster administrators to enforce `ProviderProfile` without the need for developers creating `Shoot` clusters to know which `ProviderProfiles` are available for them.

Reply:
With this being decoupled, admins can deploy more sophisticated Policy solutions such as [OpenPolicyAgent](https://www.openpolicyagent.org/) which can be used for both mutation and validation.

Reply:
> I would use shootSelector in the ProviderProfile to match Shoots.
> without the need for developers creating Shoot clusters to know which ProviderProfiles are available for them.

I don't get it. They need to know the labels that would be part of the mentioned `shootSelector`. Hence, they would first need to lookup all provider profiles, identify an adequate one, and then lookup its labels. And after that they need to label their shoot accordingly (btw, dashboard has to do the same). IMHO way too different at the moment. We don't have concrete use-cases for OPA and, to go forward (some providers are waiting for our new APIs to be available such that they can use Gardener with their already written extension controllers) I'd keep the proposal/behaviour as it is today (unless there are no other proposals).
I'm open to have a separate discussing about this and if we want to get this behaviour we probably can introduce it later without difficulties.

Reply:
For me the strongest argument to deny ShootSelector idea is shoot has hard dependency on providerProfile. IMHO, in Kubernetes world labels is optional and use for grouping/filtering resource. But here we cannot deploy shoot without knowing its providerProfile.  i.e if user creates shoot without any label, gardener won't have any idea which infra to use for interpreting secret and creating different ressources. You cannot even have default infra/providerProfile definition in such case.

I don't have idea about OpenPolicyAgent so won't argu about that scenario. But i'm sure irrespective of it shootSelector on ProviderProfile is not good idea.

Reply:
> is shoot has hard dependency on providerProfile.

And this is something which I would **REAAAAALY** like get rid of - it should be 100% possible to create a `Shoot` object without the need of `(Cloud)Profile` - anything in the `(Cloud)Profile` should be used **only** during the admission phase.

Reply:
I don't think that this can happen in the first `v1alpha1` version of this API. This needs a thorough design discussion. Let's please focus on the extensibility needs for now. Such drastic design changes could be addressed in a future version.

Reply:
> I don't think that this can happen in the first v1alpha1 version of this API. This needs a thorough design discussion. Let's please focus on the extensibility needs for now. 

Why it can't happen in the first version? Focusing on the just a small little thing, without looking at the big picture will only make things worse in the future.

> Such drastic design changes could be addressed in a future version. 

It cannot be addressed as there is backwards compatibility in question.

Reply:
> Focusing on the just a small little thing, without looking at the big picture will only make things worse in the future.

It's not a little thing. Extensibility is huge and involves a lot of risks/dangers that may affect stability of Gardener and the clusters. It's hard enough and a lot of coordination is required already. Doing too many things at once will make it worse and will break it.

> It cannot be addressed as there is backwards compatibility in question.

We already have this need now. All the new APIs have to be backwards compatible and proper conversions need to happen.

Reply:
> We already have this need now. All the new APIs have to be backwards compatible and proper conversions need to happen.

New resources in a new group have no obligation to be backwards compatible in any way if they are not backed by a the same resource storage.

Isn't it better to evolve and have both APIs working side-by side and migrate existing objects to those new API?

Reply:
> New resources in a new group have no obligation to be backwards compatible in any way if they are not backed by a the same resource storage.

But they are backed by the same resource storage.
Either way, even if this "evolved API" you are aspiring is better, it needs more time for planning, coordination, and implementation.
Now, we are focusing on #308 and need to find a solution for the people that want to integrate into Gardener with their (already written) extension controllers. They are waiting for us (there are at least three different parties we know of).

It's not only that we have to serve a new API that allows to dynamically plugin provider specifics (instead of the hard-coded AWS/Azure/... structs today), no, we also need to adapt the implementation of the controller-manager and scheduler (otherwise the new APIs are useless for providers that didn't exist previously).
If we would include such even larger changes like "shoots don't reference a cloudprofile anymore" or "separate resource for maintenance" etc. also our other tooling like dashboard, gardenctl, ... has more implementations/adaption efforts. In the end, everything will slow down and we won't be ready by end of this year.

Also, it's not clear how such change you are proposing works end-to-end. IMHO this requires a separate document and a discussion with more people. I really don't want to do such change in a hurry, especially not combined with the already expensive adaptions for extensibility as solely.